### PR TITLE
Tag table with backwards compatibility

### LINKED ISSUE #67: Feature: Store Tags in separate table rather than inline
**Is your feature request related to a problem? Please describe.**

Users would like the ability to use a separate table for tags with SQL Persistence, as described in https://github.com/akkadotnet/akka.net/issues/5296 . 

Additionally, Existing users of Akka.Persistence.Linq2Db would like to ensure they continue to maintain fast read and write performance from their existing journals.

It would be extra great, if we could provide users a way to migrate with zero downtime (if more effort.)

**Describe the solution you'd like**

We can leverage Linq2Db and Akka.Persistence.Linq2Db's pipeline to provide a solution. 

By providing 3 Read and Write modes, we can allow for a 'seamless' deployment, ideal state would be:

 - Writers and readers are in 'old' mode.
 - Deploy code for writers and readers in 'migration' mode. Writers will write both the comma separated tags and to the tag table, Readers will check for tags from both.
 - Run utility or DB Script to put old tags into the tag table.
 - Deploy code for writers and readers in 'new' mode.
 - DBAs can clear out tag column if needed.

### PR DESCRIPTION
**Pull request by to11mtm**
Fixes #67 

(Alternative solution to https://github.com/akkadotnet/akka.net/issues/5296)

## Changes

This is an implementation of Tag tables with a form of backwards compatibility.

- Added `TagWriteMode` enum, with 3 modes of operation.
  - applied logic to `ByteArrayJournalSerializer` so that tags may be applied based on setting
  - Added logic to `BaseByteArrayJournalDao` to switch behavior based on mode of operation
- Added logic to `BaseByteArrayJournalDao` to interleave batch-untagged and per-row-tagged writes when writing to tag table

- Added `TagReadMode` enum, with 3 modes of operation.
  - Added logic to `BaseByteReadArrayJournalDao` to respect enumeration
    - Queries running off tags will check tag table and/or main table based on setting
    - Filling of tags for stream pipeline will respect setting

## Todo:

 * [ ] Some cleanup in the Read Journal and elsewhere
 * [ ] Config bits (defaults, reference, table/column names)
 * [ ] Create table bits
 * [ ] Tests around new functionality
   - Ideally, we should have tests around migration scenarios. 
 * [ ] Update deletion logic to delete tags in tag table
   - *Note:* a combination of 'delete-compatibility-mode' and tag-tables may make deletions very painful, since you'd be locking across 3 tables.
 * [ ] Update 'update' functionality so that it will update tag table if needed.
   - the update functionality is a non-standard addition present from persistence-jdbc.

## Not in scope of this PR:

 - Migration routines/scripts. I think a helper application similar to what has been proposed for indexes would be a good idea, but this is likely a big enough PR as-is.

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [ ] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [ ] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [ ] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [ ] Design discussion issue #
* [ ] Changes in public API reviewed, if any.
* [ ] I have added website documentation for this feature.

### Latest `dev` Benchmarks 

Include data from the [relevant benchmark](https://getakka.net/community/contributing/index.html#improve-performance) prior to this change here.

### This PR's Benchmarks

Include data from after this change here.

### CODE REVIEW

File: src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs [hunk 1]
```diff
@@ -148,6 +154,52 @@ public AkkaPersistenceDataConnectionFactory(IProviderConfig<SnapshotTableConfigu
...
+            if (config.TableConfig.UseEventManifestColumn)
+            {
+                journalRowBuilder.Member(r => r.eventManifest)
+                    .IsColumn().HasLength(64);
+            }
+            else
+            {
+                journalRowBuilder.Member(r => r.eventManifest)
+                    .IsNotColumn();   
+            }
```
Comment (to11mtm) [inline src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs]: Basically, this switch should control whether the 'fixed' behavior for eventManifest is used.

If it's not mapped, the property will come back null and the other logic works as it should, and is ignored on inserts.

File: src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs [hunk 2]
```diff
@@ -23,6 +24,37 @@
 
 namespace Akka.Persistence.Sql.Linq2Db.Journal.DAO
 {
+    public class SequentialUUIDGenerator
```
Comment (to11mtm) [inline src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs]: This algo is borrowed from EF Core, and should be attributed such and re-commented for our use.

The expectation is that we should be able to generate Guid-esque structures with some 'predictability' in a range of output.

This function's use may require fine tuning to produce optimal results.

File: src/Akka.Persistence.Sql.Linq2Db/Query/Dao/BaseByteReadArrayJournalDAO.cs [hunk 3]
```diff
@@ -100,53 +103,271 @@ private static int MaxTake(long max)
...
+                                    r.ordering <= input.maxOffset)
+                                .Take(input.maxTake).ToListAsync();
+                            await addTagDataFromTagTable(mainRows, conn);
+                            return mainRows;
+                        }
+                    }).Via(perfectlyMatchTag(tag, separator))
+                .Via(deserializeFlow);
+        }
+
+        private Expression<Func<JournalRow, bool>> eventsByTagMigrationPredicate(DataConnection conn, string tagVal)
```
Comment (to11mtm) [inline src/Akka.Persistence.Sql.Linq2Db/Query/Dao/BaseByteReadArrayJournalDAO.cs]: This had to be expanded into it's own method because of the compiler version used in build.

File: src/Akka.Persistence.Sql.Linq2Db/Query/Dao/BaseByteReadArrayJournalDAO.cs [hunk 4]
```diff
@@ -100,53 +103,271 @@ private static int MaxTake(long max)
...
+                            return mainRows;
+                        }
+                    })
+                //We still PerfectlyMatchTag here
+                //Because DB Collation :)
+                .Via(perfectlyMatchTag(tag, separator))
                 .Via(deserializeFlow);
+        }
 
+        private Expression<Func<JournalRow, JournalTagRow, bool>> EventsByTagOnlyJoinPredicate
```
Comment (to11mtm) [inline src/Akka.Persistence.Sql.Linq2Db/Query/Dao/BaseByteReadArrayJournalDAO.cs:219]: This had to be expanded as such due to build server compiler not being able to infer expressions being returned from a ternary.

**Comment by Aaronontheweb** [comment]
Going to review this, as I am very interested in having this functionality for a project.

File: src/Akka.Persistence.Sql.Linq2Db/Config/BaseByteArrayJournalDaoConfig.cs [hunk 5]
```diff
@@ -43,6 +44,6 @@ public BaseByteArrayJournalDaoConfig(Configuration.Config config)
         public int BufferSize { get; protected set; }
         
         public bool SqlCommonCompatibilityMode { get; protected set; }
-        
+        public int DbRoundTripTagBatchSize { get; set; }
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Config/BaseByteArrayJournalDaoConfig.cs]: How does this get used exactly?

File: src/Akka.Persistence.Sql.Linq2Db/Config/JournalTableConfig.cs [hunk 6]
```diff
@@ -3,6 +3,19 @@
 
 namespace Akka.Persistence.Sql.Linq2Db.Config
 {
+    [Flags]
+    public enum TagWriteMode
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Config/JournalTableConfig.cs:6]: Got it - this is how we can support data written in previous journal implementations.

File: src/Akka.Persistence.Sql.Linq2Db/Config/JournalTableConfig.cs [hunk 7]
```diff
@@ -13,6 +26,12 @@ public class JournalTableConfig
         public string MetadataTableName { get; protected set; }
         public MetadataTableColumnNames MetadataColumnNames { get; protected set; }
         public bool WarnOnAutoInitializeFail { get; }
+        
+        public TagWriteMode TagWriteMode { get; }
+        public TagTableMode TagTableMode { get; }
+        public string? TagTableName { get; }
+        public bool UseEventManifestColumn { get; }
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Config/JournalTableConfig.cs]: These are all of the aforementioned compat-flags?

File: src/Akka.Persistence.Sql.Linq2Db/Config/JournalTableConfig.cs [hunk 8]
```diff
@@ -28,6 +47,31 @@ public JournalTableConfig(Configuration.Config config)
...
+                res = TagWriteMode.CommaSeparatedArrayAndTagTable;
+            }
+            else if (s.Equals("tagtableonly",
+                         StringComparison.InvariantCultureIgnoreCase))
+            {
+                res = TagWriteMode.TagTable;
+            }
+            else
+            {
+                res = TagWriteMode.CommaSeparatedArray;
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Config/JournalTableConfig.cs]: Got it, this is probably how most popular platforms have handled this.

File: src/Akka.Persistence.Sql.Linq2Db/Config/ReadJournalPluginConfig.cs [hunk 9]
```diff
@@ -7,11 +9,35 @@ public ReadJournalPluginConfig(Configuration.Config config)
...
 
         public string Dao { get; set; }
 
         public string TagSeparator { get; set; }
+        public TagReadMode TagReadMode { get; set; }
+        public TagTableMode TagTableMode { get; }
+    }
+
+    [Flags]
+    public enum TagReadMode
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Config/ReadJournalPluginConfig.cs:28]: How often will a journal have a separate read / write mode? Only during a migration?

File: src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs [hunk 10]
```diff
@@ -135,7 +135,13 @@ public AkkaPersistenceDataConnectionFactory(IProviderConfig<SnapshotTableConfigu
                 .HasColumnName(tableConfig.ColumnNames.SequenceNumber)
                 .Member(r => r.Timestamp)
                 .HasColumnName(tableConfig.ColumnNames.Created);
-
+            //We can skip writing tags the old way by ignoring the column in mapping.
+            journalRowBuilder.Member(r => r.tagArr).IsNotColumn();
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs]: LGTM

File: src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs [hunk 11]
```diff
@@ -148,6 +154,52 @@ public AkkaPersistenceDataConnectionFactory(IProviderConfig<SnapshotTableConfigu
                     .Member(r=>r.sequenceNumber).IsPrimaryKey();
             }
 
+            void SetJoinCol<T>(PropertyMappingBuilder<JournalTagRow, T> builder,
+                PropertyMappingBuilder<JournalRow, long> propertyMappingBuilder)
+            {
+                if (config.TableConfig.TagTableMode ==
+                    TagTableMode.SequentialUUID)
+                {
+                    builder.Member(r => r.JournalOrderingId)
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs]: LGTM

File: src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs (hunk 2 above)
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs]: Does https://github.com/akkadotnet/akka.net/pull/5995 help with this?

File: src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs [hunk 12]
```diff
@@ -149,34 +184,145 @@ private async Task WriteJournalRows(Seq<JournalRow> xs)
...
+        private async Task InsertWithOrderingAndBulkInsertTags(DataConnection dc,
+            Seq<JournalRow> xs)
+        {
+            var tagsToInsert = new List<JournalTagRow>(xs.Count);
+            foreach (var journalRow in xs)
+            {
+                var dbid = await dc.InsertWithInt64IdentityAsync(journalRow);
+                foreach (var s1 in journalRow.tagArr)
+                {
+                    tagsToInsert.Add(new JournalTagRow(){JournalOrderingId = dbid, TagValue = s1});
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs]: LGTM

**Review by Aaronontheweb — COMMENTED** [review]
Have a partial review for you @to11mtm - have about half the files to look at still.

**Comment by ismaelhamed** [comment]
Is this a different approach than [#467](https://github.com/akka/akka-persistence-jdbc/pull/467)?

**Comment by to11mtm** [comment]
> Is this a different approach than [#467](https://github.com/akka/akka-persistence-jdbc/pull/467)?

Yes and not quite:

Yes:

 - If you are running based off Ordering, it is close to the same approach as 467, with the added kindness of having migration modes. (The migration modes add a good bit of complexity here and are where much of the deviation occurs.)

Not Quite:

 - We aren't subclassing the way JDBC does to handle the different behavior modes. Given we are trying to provide migration modes, it would have meant 3 classes and a more complex migration process.
 - We are adding the option of using some form of SequentialUUID for joining to the tag table rather than orderingID.
   - This is a mode intended for scenarios where high write throughput with tags is desired, at the expense of additional storage and possibly (likely) slower reads.
   - This may not be a good idea, or may require additional optimization
 - We may or may not be doing the Inserts the same way SlickDB does. Based on the initial PRs for Tag tables in persistence-jdbc, I'm -guessing- their approach was similar to ours (where you have to insert each row one at a time to get IDs). Assumed this is still true based on them having split logic similar to this OR, where if there are no tags they try to do a simpler write.
   - To solve this, Linq2Db will need to have multirow insert-with-return support, Or https://github.com/linq2db/linq2db/issues/2960 needs to be resolved.

File: src/Akka.Persistence.Sql.Linq2Db/Config/BaseByteArrayJournalDaoConfig.cs (hunk 5 above)
Comment (to11mtm) [reply src/Akka.Persistence.Sql.Linq2Db/Config/BaseByteArrayJournalDaoConfig.cs]: I thought I left comments, IDK where then went, so just in case...

 - This controls the max number of Tags we will round-trip in a batch.
   - Tags are small compared to message payloads in most cases, so it's OK to do more tags vs Rows in a round trip I would think.

File: src/Akka.Persistence.Sql.Linq2Db/Config/ReadJournalPluginConfig.cs (hunk 9 above)
Comment (to11mtm) [reply src/Akka.Persistence.Sql.Linq2Db/Config/ReadJournalPluginConfig.cs:28]: > Only during a migration?

I would hope so. The 'both' modes are (probably) not at all performant and you'd only want to use if you were in a 'We must have zero downtime' situation and wanted to migrate.

The only other scenario I could think of is if you have some other 'things' that are doing special stuff against your existing tag column. (i.e. querying against the old tag column in some form of reporting/diagnostic query/function/sproc.) In that case they'd have to migrate that 'thing' as well.

File: src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs (hunk 2 above)
Comment (to11mtm) [reply src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs]: probably!

File: src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs (hunk 12 above)
Comment (to11mtm) [reply src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs]: @Aaronontheweb TBH we could really use more tests around things, but I have never used tags so I have a hard time coming up with scenarios.

**Comment by Aaronontheweb** [comment]
@to11mtm I'll review this again this week. Had it on my daily to-do list for a while but would really like to see it merged and Akka.Persistence.Linq2Db become the new standard ;)

File: src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs (hunk 1 above)
Comment (Aaronontheweb) [reply src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs]: LGTM

File: src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs (hunk 2 above)
Comment (Aaronontheweb) [reply src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs]: Can we use https://github.com/akkadotnet/akka.net/pull/5995 for this instead? I'd prefer to keep 1 implementation floating around, but if that approach doesn't make sense then no worries.

**Review by Aaronontheweb — APPROVED** [review]
LGTM - however, for testing this prior to going into production I'd like to propose the following:

1. Create 3 databases: Akka.Persistence.SqlServer, Postgres, and MySQL with pre-populated data using tags - all created from the same application that uses the same persistent queries. Save those DB files and commit them into a repo somewhere.
2. Build a docker image that mounts and restores those databases.
3. Run an application that uses Akka.Persistence.Linq2Db's implementation of the journal and see if we can re-run the existing queries and produce the same results against all 3.
4. Re-run a version of this using custom object serialization, just to ensure that is also supported in the migration.

That would pass the snuff test for me to a degree where I'd feel comfortable deprecating Akka.Persistence.Sql.Common and its subsequent implementations. Our team at Petabridge can help with this. What do you think @to11mtm @ismaelhamed ?

**Comment by Aaronontheweb** [comment]
Build is failing due to NuGet.org errors at the moment.

**Comment by Aaronontheweb** [comment]
@to11mtm what do you think of my review comment above? Should we mark this as "ready" and merge it in, with a plan to complete those above steps with help from our team?

**Comment by to11mtm** [comment]
> LGTM - however, for testing this prior to going into production I'd like to propose the following:

It's not done. That TODO list at the top for starters.

> 
>     1. Create 3 databases: Akka.Persistence.SqlServer, Postgres, and MySQL with pre-populated data using tags - all created from the same application that uses the same persistent queries. Save those DB files and commit them into a repo somewhere.
> 

You can't do MySQL until https://github.com/akkadotnet/Akka.Persistence.Linq2Db/issues/3 is completed.

>     2. Build a docker image that mounts and restores those databases.
> 
>     3. Run an application that uses Akka.Persistence.Linq2Db's implementation of the journal and see if we can re-run the existing queries and produce the same results against all 3.
> 
>     4. Re-run a version of this using custom object serialization, just to ensure that is also supported in the migration.
> 
> 
> That would pass the snuff test for me to a degree where I'd feel comfortable deprecating Akka.Persistence.Sql.Common and its subsequent implementations.

The other filed issues should be considered before deprecating Sql.Common, timestamp generator is one that comes to mind.

> Our team at Petabridge can help with this. What do you think @to11mtm @ismaelhamed ?

As I've mentioned in other conversations this is not a task I've had the bandwidth to take over the finish line, both in scope of the TODOs (bandwidth) as well as more full testing (I'm not familiar enough with query to come up with good tests). I can answer questions and advise but as it stands the bus factor on this is too low.

As another side note, originally we hoped to deprecate sql.common earlier. As it stands, we risk a large influx of support issues and will risk more noise between compatibility issues of switching from sql.common vs whatever set of compatibility flags they choose (i.e. eventmanifest but no tag table, compatibility modes on different DBs... delete surprises...) something to consider.

tl;dr- TODOs need to be done before this gets merged, until then the tests you describe give a false sense this is done.

**Comment by Aaronontheweb** [comment]
@Arkatufus you should move your commits onto its own PR and try to leave @to11mtm 's work here intact, for the sake of making it easy to compare the difference between the two separately.

File: src/Akka.Linq2Db.Sandbox/Akka.Linq2Db.Sandbox.csproj [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+    <PropertyGroup>
+        <TargetFramework>netcoreapp3.1</TargetFramework>
+        <Nullable>enable</Nullable>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <PackageReference Include="Akka.Streams" Version="1.4.21" />
+      <PackageReference Include="linq2db" Version="3.4.1" />
+      <PackageReference Include="Reactive.Streams" Version="1.0.2" />
```
Comment (Aaronontheweb) [inline src/Akka.Linq2Db.Sandbox/Akka.Linq2Db.Sandbox.csproj:11]: No need for an explicit Reactive.Streams reference here

File: src/Akka.Persistence.Linq2Db.Compatibility.Tests/Akka.Persistence.Linq2Db.Compatibility.Tests.csproj [hunk 14]
```diff
@@ -9,6 +9,7 @@
     </PropertyGroup>
 
     <ItemGroup>
+        <PackageReference Include="Akka.TestKit" Version="$(AkkaVersion)" />
         <PackageReference Include="Akka" Version="$(AkkaVersion)" />
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Linq2Db.Compatibility.Tests/Akka.Persistence.Linq2Db.Compatibility.Tests.csproj:13]: Can remove explicit `Akka` reference if we're referencing the TestKit

File: src/Akka.Persistence.Sql.Linq2Db/Config/BaseByteArrayJournalDaoConfig.cs [hunk 15]
```diff
@@ -9,6 +9,7 @@ public BaseByteArrayJournalDaoConfig(Configuration.Config config)
             BufferSize = config.GetInt("buffer-size", 5000);
             BatchSize = config.GetInt("batch-size", 100);
             DbRoundTripBatchSize = config.GetInt("db-round-trip-max-batch-size", 1000);
+            DbRoundTripTagBatchSize = config.GetInt("db-round-trip-max-tag-batch-size", 1000);
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Config/BaseByteArrayJournalDaoConfig.cs:12]: LGTM

File: src/Akka.Persistence.Sql.Linq2Db/Config/TagTableConfig.cs [hunk 16]
```diff
@@ -0,0 +1,41 @@
+﻿// -----------------------------------------------------------------------
+//  <copyright file="TagTableConfig.cs" company="Akka.NET Project">
+//      Copyright (C) 2009-2023 Lightbend Inc. <http://www.lightbend.com>
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Config/TagTableConfig.cs:3]: Can get rid of Lightbend attribution here.

File: src/Akka.Persistence.Sql.Linq2Db/Config/TagTableColumnNames.cs [hunk 17]
```diff
@@ -0,0 +1,43 @@
+﻿// -----------------------------------------------------------------------
+//  <copyright file="TagTableColumnNames.cs" company="Akka.NET Project">
+//      Copyright (C) 2009-2023 Lightbend Inc. <http://www.lightbend.com>
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Config/TagTableColumnNames.cs:3]: Same here

File: src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs [hunk 18]
```diff
@@ -140,6 +140,13 @@ public AkkaPersistenceDataConnectionFactory(IProviderConfig<SnapshotTableConfigu
                 .Member(r => r.Timestamp)
                 .HasColumnName(columnNames.Created);
 
+            //We can skip writing tags the old way by ignoring the column in mapping.
+            journalRowBuilder.Member(r => r.TagArr).IsNotColumn();
+            if (tableConfig.TagWriteMode == TagWriteMode.TagTable)
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Db/AkkaPersistenceDataConnectionFactory.cs:145]: Based on @to11mtm 's suggestion, we're going to need to be able to support multiple tag write modes. Can we address this in a subsequent PR?

File: src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs [hunk 19]
```diff
@@ -15,12 +15,44 @@
...
 using Akka.Streams.Dsl;
 using LanguageExt;
-using LinqToDB;
+using LinqToDB;using LinqToDB.Common;
 using LinqToDB.Data;
 using static LanguageExt.Prelude;
 
 namespace Akka.Persistence.Sql.Linq2Db.Journal.Dao
 {
+    public class SequentialUuidGenerator
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/BaseByteArrayJournalDao.cs:24]: Can make this `static`

File: src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/ByteArrayJournalSerializer.cs [hunk 20]
```diff
@@ -18,13 +19,15 @@ public sealed class ByteArrayJournalSerializer : FlowPersistentReprSerializer<Jo
         private readonly string _separator;
         private readonly IProviderConfig<JournalTableConfig> _journalConfig;
         private readonly string[] _separatorArray;
+        private readonly TagWriteMode _tagWriteMode;
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Journal/DAO/ByteArrayJournalSerializer.cs:22]: Consideration for multiple write modes here as well

File: src/Akka.Persistence.Sql.Linq2Db/Journal/Types/JournalRow.cs [hunk 21]
```diff
@@ -1,8 +1,16 @@
...
 
 namespace Akka.Persistence.Sql.Linq2Db.Journal.Types
 {
+    public sealed class JournalTagRow
+    {
+        public long JournalOrderingId { get; set; }
+        
+        public string TagValue { get; set; }
+        
+        public Guid WriteUuid { get; set; }
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/Journal/Types/JournalRow.cs:11]: No need to include `WriteUuid` on the column - OrderingId should be good enough if it's unique.

File: src/Akka.Persistence.Sql.Linq2Db/persistence.conf [hunk 22]
```diff
@@ -129,6 +131,15 @@
             sequence-number = sequence_number
           }
         }
+        
+        tag {
+          table-name = "tags"
+          columns {
+            ordering-id = ordering_id
+            tag-value = tag
+            writer-uuid = writer_uuid
```
Comment (Aaronontheweb) [inline src/Akka.Persistence.Sql.Linq2Db/persistence.conf:140]: Can remove this

**Review by Aaronontheweb — APPROVED** [review]
Approving and merging this because it's a blocker and I don't want to review another 90+ file PR again, but there are some issues that need to be addressed in the comments. Let's deal with those in separate PRs.
